### PR TITLE
Multiple recursive definitions

### PR DESCRIPTION
**Pull request by andrejbauer**
* A more uniform notation for binders. In a binder we can write either
  a variable `x` without a typing annotation, or `(x_1 .. x_n : A)` and
  nothing else:

    * `lambda x y z, e` is ok
    * `lambda x (y : A) z, e` is ok
    * `lambda x (+) z, e` is ok
    * `lambda x ((+) : A) z, e` is ok
    * `lambda (x y : A) z, e` is ok
    * `lambda x (y) z, e` is not ok
    * `lambda x y : A, e` is not ok

* mutually recursive function definitions:

    let rec f x = c1
    and g y = c2
    and h z = c2
    in c

  and similarly toplevel `let rec`.

* parallel top-level let (not really needed but is there for
  orthogonality):

    let x1 = c1
    and x2 = c2
    and x3 = c3

* the old `rec` construct is gone

### CODE REVIEW

File: src/nucleus/value.ml [hunk 1]
```diff
@@ -1,5 +1,9 @@
 (** Runtime values and results *)
 
+type ('a,'b) sum =
```
Comment (SkySkimmer) [inline src/nucleus/value.ml]: This moved to Tt

**Comment by SkySkimmer** [comment]
>`lambda x y : A, e` is not ok
Why not?

**Comment by SkySkimmer** [comment]
Can you rebase this and #221 on master? Or cherry-pick since both only have one real commit.

**Comment by andrejbauer** [comment]
`lambda x y : A` is a major pain for parsing because you have to look waay ahead to see if a `:` is coming. I'll fix it one day, but it'll make the parser ugly(er).

**Comment by andrejbauer** [comment]
I'll cherry-pick.

**Pull request closed without merging**